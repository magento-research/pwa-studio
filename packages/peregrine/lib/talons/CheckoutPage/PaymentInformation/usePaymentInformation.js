import { useState, useCallback } from 'react';
import { useQuery } from '@apollo/react-hooks';

import { useAppContext } from '../../../context/app';
import { useCartContext } from '../../../context/cart';

/**
 *
 * @param {Function} props.onSave callback to be called when user clicks review order button
 * @param {DocumentNode} props.operations.queries.getSelectedPaymentMethodQuery query to get the selected payment method value from cache
 * @param {DocumentNode} props.operations.queries.getPaymentNonceQuery query to get the payment nonce from cache
 *
 * @returns {
 *   doneEditing: Boolean,
 *   shouldRequestPaymentNonce: Boolean,
 *   isEditModalHidden: Boolean,
 *   selectedPaymentMethod: String,
 *   paymentNonce: {
 *      nonce: String,
 *      type: String,
 *      description: String,
 *      details: {
 *          cardType: String,
 *          lastFour: String,
 *          lastTwo: String
 *      },
 *      binData: {
 *          prepaid: String,
 *          healthcare: String,
 *          debit: String,
 *          durbinRegulated: String,
 *          commercial: String,
 *          payroll: String,
 *          issuingBank: String,
 *          countryOfIssuance: String,
 *          productId: String
 *      }
 *   },
 *   handleReviewOrder: Function,
 *   showEditModal: Function,
 *   hideEditModal: Function
 *
 * }
 */
export const usePaymentInformation = props => {
    const { onSave, operations } = props;
    const {
        queries: { getSelectedPaymentMethodQuery, getPaymentNonceQuery }
    } = operations;

    /**
     * Definitions
     */

    const [shouldRequestPaymentNonce, setShouldRequestPaymentNonce] = useState(
        false
    );
    const [isEditModalHidden, setIsEditModalHidden] = useState(true);
    const [, { toggleDrawer, closeDrawer }] = useAppContext();
    const [{ cartId }] = useCartContext();

    /**
     * Query Fetches
     */

    const { data: selectedPaymentMethodData } = useQuery(
        getSelectedPaymentMethodQuery,
        {
            variables: {
                cartId
            }
        }
    );

    const { data: paymentNonceData } = useQuery(getPaymentNonceQuery, {
        variables: {
            cartId
        }
    });

    const selectedPaymentMethod = selectedPaymentMethodData
        ? selectedPaymentMethodData.cart.selectedPaymentMethod
        : null;

    const paymentNonce = paymentNonceData
        ? paymentNonceData.cart.paymentNonce
        : null;

    /**
     * Helper Functions
     */

    const handleReviewOrder = useCallback(() => {
        setShouldRequestPaymentNonce(true);
        if (onSave) {
            onSave();
        }
    }, [onSave]);

    const showEditModal = useCallback(() => {
        toggleDrawer('edit.payment');
        setIsEditModalHidden(false);
    }, [setIsEditModalHidden, toggleDrawer]);

    const hideEditModal = useCallback(() => {
        setIsEditModalHidden(true);
        closeDrawer('edit.payment');
    }, [setIsEditModalHidden, closeDrawer]);

    return {
        doneEditing: !!paymentNonce,
        handleReviewOrder,
        shouldRequestPaymentNonce,
        selectedPaymentMethod,
        paymentNonce,
        isEditModalHidden,
        showEditModal,
        hideEditModal
    };
};
