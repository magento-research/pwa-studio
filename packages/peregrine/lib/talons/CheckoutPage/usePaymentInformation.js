

import { useFieldState } from 'informed';

const usePaymentInformation = () => {
    const selectedPaymentMethod = useFieldState('testradiobuttons');

    return {
        selectedPaymentMethod
    };
};

export default usePaymentInformation;
