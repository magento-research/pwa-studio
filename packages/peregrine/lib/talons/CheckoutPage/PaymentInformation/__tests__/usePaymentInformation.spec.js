import React from 'react';
import { useQuery } from '@apollo/react-hooks';

import { usePaymentInformation } from '../usePaymentInformation';
import createTestInstance from '../../../../util/createTestInstance';

jest.mock('../../../../context/cart', () => ({
    useCartContext: jest.fn().mockReturnValue([{ cartId: '123' }])
}));

jest.mock('../../../../context/app', () => ({
    useAppContext: jest
        .fn()
        .mockReturnValue([
            {},
            { toggleDrawer: () => {}, closeDrawer: () => {} }
        ])
}));

jest.mock('@apollo/react-hooks', () => {
    return { useQuery: jest.fn() };
});

jest.mock('informed', () => {
    return {
        useFieldState: jest.fn().mockReturnValue({ value: 'braintree' })
    };
});

const Component = props => {
    const talonProps = usePaymentInformation(props);

    return <i talonProps={talonProps} />;
};

const getTalonProps = props => {
    const tree = createTestInstance(<Component {...props} />);
    const { talonProps } = tree.root.findByType('i').props;

    const update = newProps => {
        tree.update(<Component {...{ ...props, ...newProps }} />);

        return tree.root.findByType('i').props;
    };

    return { talonProps, tree, update };
};

const getCheckoutStepQuery = 'getCheckoutStepQuery';
const queries = {
    getCheckoutStepQuery
};

const getCheckoutStep = jest.fn().mockReturnValue({
    data: { cart: { checkoutStep: 3 } }
});

beforeAll(() => {
    useQuery.mockImplementation(query => {
        if (query === getCheckoutStepQuery) {
            return getCheckoutStep();
        } else {
            return { data: {} };
        }
    });
});

test('Should return correct shape', () => {
    const { talonProps } = getTalonProps({ queries });

    expect(talonProps).toMatchSnapshot();
});

test('doneEditing should be true if checkoutStep is a greater than 3', () => {
    getCheckoutStep.mockReturnValueOnce({
        data: { cart: { checkoutStep: 3 } }
    });

    const { talonProps, update } = getTalonProps({ queries });

    expect(talonProps.doneEditing).toBeFalsy();

    getCheckoutStep.mockReturnValueOnce({
        data: {
            cart: {
                checkoutStep: 4
            }
        }
    });

    const { talonProps: newTalonProps } = update();

    expect(newTalonProps.doneEditing).toBeTruthy();
});

test('shouldRequestPaymentNonce should be set to true when handleReviewOrder is called', () => {
    const onSave = jest.fn();
    const { talonProps, update } = getTalonProps({ queries, onSave });

    expect(talonProps.shouldRequestPaymentNonce).toBeFalsy();
    expect(onSave).not.toHaveBeenCalled();

    talonProps.handleReviewOrder();

    const { talonProps: newTalonProps } = update();

    expect(newTalonProps.shouldRequestPaymentNonce).toBeTruthy();
    expect(onSave).not.toHaveBeenCalled();
});

test('onSave should be called when handlePaymentSuccess is called', () => {
    // TODO
    const onSave = jest.fn();
    const { talonProps, update } = getTalonProps({ queries, onSave });

    expect(onSave).not.toHaveBeenCalled();

    talonProps.handlePaymentSuccess();

    update();

    expect(onSave).toHaveBeenCalled();
});

test('shouldRequestPaymentNonce should be set to false when handlePaymentError is called', () => {
    const onSave = jest.fn();
    const { talonProps, update } = getTalonProps({ queries, onSave });

    expect(talonProps.shouldRequestPaymentNonce).toBeFalsy();
    expect(onSave).not.toHaveBeenCalled();

    talonProps.handleReviewOrder();

    const { talonProps: step1TalonProps } = update();

    expect(step1TalonProps.shouldRequestPaymentNonce).toBeTruthy();
    expect(onSave).not.toHaveBeenCalled();

    talonProps.handlePaymentError();

    const { talonProps: step2TalonProps } = update();

    expect(step2TalonProps.shouldRequestPaymentNonce).toBeFalsy();
    expect(onSave).not.toHaveBeenCalled();
});

test('hideEditModal and showEditModal functions should toggle isEditModalHidden flag', () => {
    const { talonProps, update } = getTalonProps({ queries });

    expect(talonProps.isEditModalHidden).toBeTruthy();

    talonProps.showEditModal();

    const { talonProps: step1Props } = update();

    expect(step1Props.isEditModalHidden).toBeFalsy();

    talonProps.hideEditModal();

    const { talonProps: step2Props } = update();

    expect(step2Props.isEditModalHidden).toBeTruthy();
});
