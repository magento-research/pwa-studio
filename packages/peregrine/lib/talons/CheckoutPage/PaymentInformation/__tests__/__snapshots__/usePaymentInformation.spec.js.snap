// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should return correct shape 1`] = `
Object {
  "doneEditing": true,
  "handlePaymentError": [Function],
  "handlePaymentSuccess": [Function],
  "handleReviewOrder": [Function],
  "hideEditModal": [Function],
  "isEditModalHidden": true,
  "paymentNonce": Object {},
  "selectedPaymentMethod": "creditCard",
  "shouldRequestPaymentNonce": false,
  "showEditModal": [Function],
}
`;
